environment.go

```diff
@@ -16,10 +16,120 @@ package esc
 
 import (
 	"encoding/json"
+	"errors"
+	"fmt"
 
 	"github.com/pulumi/esc/schema"
 )
 
+const AnonymousEnvironmentName = "<yaml>"
+
+type ExecContext struct {
+	rootEvironment string
+	values         map[string]Value
+}
+
+type copier struct {
+	memo map[*Value]*Value
+}
+
+func newCopier() copier {
+	return copier{memo: map[*Value]*Value{}}
```

Comment:
I can merge this to reuse the copier between eval/val and this

https://github.com/glena/esc/compare/glena:esc:german/root-current-env-context...glena:esc:german/root-current-env-context-copy?expand=1

Reply:
I think it's fine to have two implementations for now: using the generic copier feels a bit risky to me given the sensitivity of the `*value` graph to proper memoized copies.

FWIW, the `*Value`s don't have the same restrictions, so we could also simplify this by removing the memoization.